policy-controller/k8s/status/src/index.rs

```diff
@@ -47,21 +54,50 @@ pub struct Update {
 }
 
 impl Controller {
-    pub fn new(client: k8s::Client, updates: UnboundedReceiver<Update>) -> Self {
-        Self { client, updates }
+    pub fn new(
+        claims: Receiver<Arc<Claim>>,
+        client: k8s::Client,
+        name: String,
+        updates: UnboundedReceiver<Update>,
+    ) -> Self {
+        Self {
+            claims,
+            client,
+            name,
+            updates,
+            leader: false,
+        }
     }
 
-    pub async fn process_updates(mut self) {
-        let patch_params = k8s::PatchParams::apply("policy.linkerd.io");
+    pub async fn run(mut self) {
+        let patch_params = k8s::PatchParams::apply(POLICY_API_GROUP);
 
-        // todo: If an update fails we should figure out a requeueing strategy
-        while let Some(Update { id, patch }) = self.updates.recv().await {
-            let api =
-                k8s::Api::<k8s::policy::HttpRoute>::namespaced(self.client.clone(), &id.namespace);
+        // Select between the status-controller lease claim changing and
+        // receiving updates from the index. If the lease claim changes, then
+        // check if we are now the leader. If so, we should apply the patches
+        // received; otherwise, we should drain the updates queue but not
+        // apply any patches since another status controller is responsible
+        // for that.
+        loop {
+            tokio::select! {
+                biased;
+                _ = self.claims.changed() => {
+                    let claim = self.claims.borrow_and_update();
+                    self.leader =  claim.is_current_for(&self.name)
+                }
+                Some(Update { id, patch}) = self.updates.recv() => {
+                    // If this status controller is not the leader, it should
+                    // process through the updates queue but not actually
+                    // patch any resources.
+                    if !self.leader {
+                        continue;
+                    }
```

Comment:
take it or leave it: we *could* potentially roll the `if !self.leader` check into the `select!` condition, like this:
```suggestion
                Some(Update { id, patch}) = self.updates.recv() if self.is_leader => {
```
although i think the logic might be easier to follow leaving it separate, and the `if` adds a nice place to add a comment.

i could be convinced either way about this :woman_shrugging:

Reply:
https://github.com/linkerd/linkerd2/pull/10444/commits/d7e67a7677595b336f5c19716fe2c9f76ecba5ac. I moved the comment above the branch and don't think it's any harder to follow what's going on.